java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.qll

```diff
@@ -0,0 +1,76 @@
+import java
+import DataFlow
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.Servlets
+
+/**
+ * A concatenate expression using the string `forward:` on the left.
+ *
+ * E.g: `"forward:" + url`
+ */
+class ForwardBuilderExpr extends AddExpr {
+  ForwardBuilderExpr() {
+    this.getLeftOperand().(CompileTimeConstantExpr).getStringValue() = "forward:"
+  }
+}
+
+/**
+ * A call to `StringBuilder.append` or `StringBuffer.append` method, and the parameter value is `"forward:"`.
+ *
+ * E.g: `StringBuilder.append("forward:")`
+ */
+class ForwardAppendCall extends MethodAccess {
+  ForwardAppendCall() {
+    this.getMethod().hasName("append") and
+    this.getMethod().getDeclaringType() instanceof StringBuildingType and
+    this.getArgument(0).(CompileTimeConstantExpr).getStringValue() = "forward:"
+  }
+}
+
+abstract class UnsafeUrlForwardSink extends DataFlow::Node { }
+
+/** A Unsafe url forward sink from getRequestDispatcher method. */
+private class RequestDispatcherSink extends UnsafeUrlForwardSink {
+  RequestDispatcherSink() {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof ServletRequestGetRequestDispatcherMethod and
+      ma.getArgument(0) = this.asExpr()
+    )
+  }
+}
+
+/** A Unsafe url forward sink from spring controller method. */
+private class SpringUrlForwardSink extends UnsafeUrlForwardSink {
```

Comment:
Overlap with CWE-601/SpringUrlRedirect.ql?

Reply:
No overlap, no url jump will be triggered here, files will be read.